### ISSUE TITLE
Design tasks with deferTaskConfiguration() keep in mind

### ISSUE THREAD

**Issue by Sweetchuck** [consolidation/robo#1078]
### Expected behavior
```php
<?php

class MyCommand extends \Robo\Tasks {
    protected function foo() {
        return $this
            ->collectionBuilder()
            ->addCode(function (\Robo\State\Data $state): int {
                $state['my_dynamic_src'] = '/my_src.zip';
                $state['my_dynamic_dst'] = '/my_dst';

                return 0;
            })
            ->addTask(
                $this
                    ->taskExtract('this_should_not_be_required.zip')
                    ->deferTaskConfiguration('fileName', 'my_dynamic_src')
                    ->deferTaskConfiguration('to', 'my_dynamic_dst')
            );
    }
}
```

### Actual behavior
Unfortunately the $fileName parameter is required in the [Task/Archive/Extract::__constructor](https://github.com/consolidation/robo/blob/1cdf254b76eaaa81672347b6d8f71afffd273444/src/Task/Archive/Extract.php#L57), and cannot be overwritten later.

## Affected tasks (and there are more)

I realized that this problem is more common.

* CleanDir
* CopyDir
* [\Robo\Task\Filesystem\DeleteDir](https://github.com/consolidation/robo/blob/3.x/src/Task/Filesystem/DeleteDir.php)
* FlattenDir
* MirrorDir
* \Robo\Task\Docker\Exec ContainerID or Result
* \Robo\Task\Docker\Stop ContainerID or Result
* \Robo\Task\Docker\Commit ContainerID or Result 
* \Robo\Task\Docker\Pull Image name
* \Robo\Task\Docker\Build path

So basically every task should be compatible with the „deferTaskConfiguration” concept. Currently there are sever task which can be configured only through the `::taskFoo(...)` method – which is provided by a trait – and the task receives the configurations by __construct() arguments, and those values has to be provided ( problem 1) and cannot be changed later (problem 2)
